### PR TITLE
Add linspaced_int_array()

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Adds support for `ìnt[] linspaced_int_array(int, int, int)` as a complement to `real[] linspaced_int_array(int, real, real)`. This was proposed by @avehtari in a discussion and I agree it would be useful.

`linspaced_int_array(5,0,8)` returns 0,2,4,6,8
`linspaced_int_array(5,0,9)` returns 0,2,4,6,8
`linspaced_int_array(5,0,11)` returns 0,2,4,6,8
`linspaced_int_array(5,0,12)` returns 0,3,6,9,12

So the behaviour is the same as for `Eigen::VectorXi v = Eigen::VectorXi::LinSpaced(K, low, high)`

## Tests

/

## Side Effects

/

## Release notes

Added linspaced_int_array.

## Checklist

- [x] Copyright holder: Rok Češnovar

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/prim/fun/linspaced_int_array.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * as (high - low) / (K - 1). For K=1, the array will contain the high value;
+ * for K=0 it returns an empty array.
+ *
+ * @param K size of the array
+ * @param low smallest value
+ * @param high largest value
+ * @return An array of size K with elements linearly spaced between
+ * low and high.
+ * @throw std::domain_error if K is negative, if high is less than low
+ * or if ().
```
Comment (bbbales2) [inline stan/math/prim/fun/linspaced_int_array.hpp:24]: `if ()`?

File: stan/math/prim/fun/linspaced_int_array.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return an array of linearly spaced integers.
+ *
+ * This produces an array from low to high (included) with integers spaced
```
Comment (bbbales2) [inline stan/math/prim/fun/linspaced_int_array.hpp:14]: included -> inclusive

File: stan/math/prim/fun/linspaced_int_array.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * @param K size of the array
+ * @param low smallest value
+ * @param high largest value
+ * @return An array of size K with elements linearly spaced between
+ * low and high.
+ * @throw std::domain_error if K is negative, if high is less than low
+ * or if ().
+ */
+inline std::vector<double> linspaced_int_array(int K, int low, int high) {
```
Comment (bbbales2) [inline stan/math/prim/fun/linspaced_int_array.hpp:26]: `std::vector<int>`

File: stan/math/prim/fun/linspaced_int_array.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  static const char* function = "linspaced_int_array";
+  check_nonnegative(function, "size", K);
+  check_greater_or_equal(function, "high", high, low);
+  if (K == 0) {
+    return {};
+  }
+
+  Eigen::VectorXi v = Eigen::VectorXi::LinSpaced(K, low, high);
```
Comment (bbbales2) [inline stan/math/prim/fun/linspaced_int_array.hpp:34]: Build the `std::vector<int>` directly.

Some variation of this I guess:
```c++
if(K == 1) {
  return { high };
}

std::vector<int> v(K);
int d = (high - low) / (K - 1);
for(size_t k = 0; k < K; ++k) {
  v[k] = k * d + low;
}
return v;
```

File: test/unit/math/prim/fun/linspaced_int_array_test.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include <stan/math/prim.hpp>
+#include <test/unit/util.hpp>
+#include <gtest/gtest.h>
+#include <limits>
+#include <vector>
+
+TEST(MathFunctions, linspaced_int_array) {
+  using stan::math::linspaced_int_array;
+  EXPECT_STD_VECTOR_EQ(linspaced_int_array(0, 1, 5), std::vector<int>({}));
+  EXPECT_STD_VECTOR_EQ(linspaced_int_array(1, 1, 5), std::vector<int>({5}));
```
Comment (bbbales2) [inline test/unit/math/prim/fun/linspaced_int_array_test.cpp:10]: Huh, that's surprising. I'd have expected the output to be `1`. `5` seems strange. Is that really what Eigen does?

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Review!

File: test/unit/math/prim/fun/linspaced_int_array_test.cpp (hunk 5 above)
Comment (bbbales2) [reply test/unit/math/prim/fun/linspaced_int_array_test.cpp:10]: Actually I guess we should do whatever the already existing linspaced stuff does.

Comment (rok-cesnovar) [reply test/unit/math/prim/fun/linspaced_int_array_test.cpp:10]: Yeah that is why its done that way.

File: stan/math/prim/fun/linspaced_int_array.hpp (hunk 4 above)
Comment (rok-cesnovar) [reply stan/math/prim/fun/linspaced_int_array.hpp:34]: Is there an advantage to that approach? This is the same impl. as the other linspaced function. But can also do that.

File: stan/math/prim/fun/linspaced_int_array.hpp (hunk 3 above)
Comment (bbbales2) [reply stan/math/prim/fun/linspaced_int_array.hpp:25]: This is still a `std::vector<double>` on return.

File: stan/math/prim/fun/linspaced_int_array.hpp (hunk 4 above)
Comment (bbbales2) [reply stan/math/prim/fun/linspaced_int_array.hpp:33]: It avoids two allocates (the local `std::vector<int>` will get returned directly by compiler optimizing returns) and the allocates and copies will be the expensive part of this.

But in the spirit of not tripping up on edge cases (that implementation I gave I think doesn't work for some of the examples I see [here](http://eigen.tuxfamily.org/dox-devel/classEigen_1_1DenseBase.html#title49)), I'll also take ```return {v.data(), v.data() + K};```

Comment (rok-cesnovar) [reply stan/math/prim/fun/linspaced_int_array.hpp:33]: Made it return {v.data(), v.data() + K};

File: stan/math/prim/fun/linspaced_int_array.hpp (hunk 3 above)
Comment (rok-cesnovar) [reply stan/math/prim/fun/linspaced_int_array.hpp]: Bah, silly me

**Comment by bbbales2** [comment]
I'm still confused on the docs on this.

This is how Eigen says it (from [here](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#aaef589c1dbd7fad93f97bd3fa1b1e768)):

> For integer scalar types, an even spacing is possible if and only if the length of the range, i.e., high-low is a scalar multiple of size-1, or if size is a scalar multiple of the number of values high-low+1 (meaning each value can be repeated the same number of time). If one of these two considions is not satisfied, then high is lowered to the largest value satisfying one of this constraint.

That last sentence is hard to parse.

**Comment by rok-cesnovar** [comment]
I think it just mean when we specify

Eigen::VectorXi::LinSpaced(5, 0, 11); it lowers 11 until its a valid linspaced vector (under the two conditions).

**Review by bbbales2 — COMMENTED** [review]
Lemme know if you approve my changes and I'll merge. This looks good to me.

**Comment by rok-cesnovar** [comment]
All good! Thanks.

**Comment by rok-cesnovar** [comment]
@bbbales2 approve?

**Review by bbbales2 — APPROVED** [review]
